### PR TITLE
merge winkernel and dummy datapath

### PR DESCRIPTION
**Pull request by ami-GS**
## Description

Kernel datapath has been independent. now merging to dummy datapath for future xdp integration

## Testing

CI

## Documentation

N/A

### CODE REVIEW

File: src/platform/datapath_raw.h [hunk 1]
```diff
@@ -474,11 +437,52 @@ typedef struct TCP_HEADER {
 #define TH_CWR 0x80
 
 #define IPV4_VERSION 4
-#define IPV6_VERSION 6
 #define IPV4_VERSION_BYTE (IPV4_VERSION << 4)
-#define IPV4_DEFAULT_VERHLEN ((IPV4_VERSION_BYTE) | (sizeof(IPV4_HEADER) / sizeof(uint32_t)))
 
 #define IP_DEFAULT_HOP_LIMIT 128
 
+#ifndef _KERNEL_MODE
```
Comment (mtfriesen) [inline src/platform/datapath_raw.h:444]: why are these defs kernel-specific?

Comment (ami-GS) [reply src/platform/datapath_raw.h:444]: These types seem to be defined somewhere for kernel mode

File: src/platform/datapath_raw.h [hunk 2]
```diff
@@ -474,11 +437,52 @@ typedef struct TCP_HEADER {
...
+    uint8_t Data[0];
+} IPV6_HEADER;
+
+#define IPV6_VERSION 6
+#define IPV4_DEFAULT_VERHLEN ((IPV4_VERSION_BYTE) | (sizeof(IPV4_HEADER) / sizeof(uint32_t)))
+
 #define ETHERNET_TYPE_IPV4 0x0008
 #define ETHERNET_TYPE_IPV6 0xdd86
+
+#endif
```
Comment (nibanks) [inline src/platform/datapath_raw.h:488]: nit: trailing newline please

File: src/platform/datapath_winkernel.c [hunk 3]
```diff
@@ -2670,14 +2460,15 @@ CxPlatSendDataAlloc(
                 ? Config->MaxPacketSize : 0;
         SendData->ClientBuffer.Length = 0;
         SendData->ClientBuffer.Buffer = NULL;
+        SendData->DatapathType = Config->Route->DatapathType = CXPLAT_DATAPATH_TYPE_USER;
```
Comment (mtfriesen) [inline src/platform/datapath_winkernel.c:2463]: `TYPE_USER` in kernel mode? This should probably be renamed.

File: src/platform/platform_internal.h [hunk 4]
```diff
@@ -191,6 +194,139 @@ typedef struct CX_PLATFORM {
...
+    CXPLAT_DATAPATH_PROC_CONTEXT ProcContexts[0];
+
+} CXPLAT_DATAPATH;
+
+#ifndef htonl
+#define htonl _byteswap_ulong
+#endif
+
+#define IS_LOOPBACK(Address) ((Address.si_family == QUIC_ADDRESS_FAMILY_INET &&                \
+                               Address.Ipv4.sin_addr.S_un.S_addr == htonl(INADDR_LOOPBACK)) || \
```
Comment (mtfriesen) [inline src/platform/platform_internal.h:326]: I'm sure this is just being refactored in this PR, but consider `IN4_IS_ADDR_LOOPBACK` to match v6 below, or even just `INET_IS_ADDR_LOOPBACK`

**Review by mtfriesen — APPROVED** [review]

File: src/platform/datapath_winkernel.c [hunk 5]
```diff
@@ -3192,55 +2983,11 @@ CxPlatSocketGetTcpStatistics(
...
-    _Inout_ CXPLAT_ROUTE* DstRoute,
-    _In_ CXPLAT_ROUTE* SrcRoute
-    )
-{
-    UNREFERENCED_PARAMETER(DstRoute);
-    UNREFERENCED_PARAMETER(SrcRoute);
-}
+    UNREFERENCED_PARAMETER(Cqe);
+    CXPLAT_DBG_ASSERT(FALSE);
+}
```
Comment (nibanks) [inline src/platform/datapath_winkernel.c:2993]: nit: need a trailing newline please

**Pull request merged**